From 429c2b8d5e1f644028980dbb8c2f0bba76f4de48 Mon Sep 17 00:00:00 2001
From: "Song, Hongyan" <hongyan.song@intel.com>
Date: Fri, 3 Aug 2018 14:45:59 +0800
Subject: [PATCH 1/2] HID: increase maximum global item tag report size to 256
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The maximum globale report size has changed from 32->...->96->128 in the past
years.  With the development usage of HID, the report_size max value 128 cannot
satisfy all requirements.

There are applications need to expose intrinsic metadata to camera stabilizing
applications such as 3DFE application. 3DFE intrinsic is designed to express
environmental information about sensor that may dynamically change while the
sensor is running (such data include noise spectral density, bias standard
deviation)

A sensor data field is SENSOR_VALUE_PAIR that consists of a PROPERTYKEY and
PROPVARIANT pair. It need to report a unique PROPERTYKEY for each data field.

Take “Noise Spectral Density” as an example, it report count will be
defined as below:

	"Size of Property key GUID(16 Byte) + property key index(4 Byte) +
	size of Noise Spectral Density value(4 Byte)"

In this case, the data report max is totally 192(24Byte), which is larger than
128, while max size 128 blocked it as illegal length. So increase the report
size to satisfy it and more demands in the future.

Signed-off-by: Song Hongyan <hongyan.song@intel.com>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>

(cherry picked from commit 71f6fa90a353605bf25c36417c9ae529ac1a9a8d)
[Reason for cherry-picking this commis:
  Reported that this is needed for Surface Go 2. Maybe also needed for
  Surface Pro 7 and Surface Laptop 3.
  Below is the dmesg log from SGO2 without this commit:
    [   19.180616] hid (null): invalid report_size 192
    [   19.183874] hid (null): invalid report_size 192
    [   19.186760] hid (null): invalid report_size 192
    [   19.189703] hid (null): invalid report_size 192
    [   19.192945] hid (null): invalid report_size 192
    [   19.195875] hid (null): invalid report_size 192
    [   19.198523] hid (null): invalid report_size 192
    [   19.201032] hid (null): invalid report_size 192
    [   19.204702] hid-generic 001F:8087:0AC2.0006: invalid report_size 192
    [   19.207506] hid-generic 001F:8087:0AC2.0006: item 0 1 1 7 parsing failed
    [   19.210194] hid-generic: probe of 001F:8087:0AC2.0006 failed with error -22
    [   19.215825] hid-generic 001F:8087:0AC3.0007: hidraw5: <UNKNOWN> HID v2.00 Device [hid-ishtp 8087:0AC3] on
    [   19.218727] hid-sensor-hub 001F:8087:0AC2.0006: invalid report_size 192
    [   19.220200] hid-sensor-hub 001F:8087:0AC2.0006: item 0 1 1 7 parsing failed
    [   19.221752] hid-sensor-hub 001F:8087:0AC2.0006: parse failed
    [   19.223354] hid-sensor-hub: probe of 001F:8087:0AC2.0006 failed with error -22
  Without this commit, reported that sensors are not working on SGO2.
  This commit is available after v4.20-rc1, so backport this into v4.19
  series.]
Signed-off-by: Tsuchiya Yuto (kitakar5525) <kitakar@gmail.com>
---
 drivers/hid/hid-core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/hid/hid-core.c b/drivers/hid/hid-core.c
index b26206fbb457e..3f6534cb97143 100644
--- a/drivers/hid/hid-core.c
+++ b/drivers/hid/hid-core.c
@@ -444,7 +444,7 @@ static int hid_parser_global(struct hid_parser *parser, struct hid_item *item)
 
 	case HID_GLOBAL_ITEM_TAG_REPORT_SIZE:
 		parser->global.report_size = item_udata(item);
-		if (parser->global.report_size > 128) {
+		if (parser->global.report_size > 256) {
 			hid_err(parser->device, "invalid report_size %d\n",
 					parser->global.report_size);
 			return -1;
-- 
2.27.0

